boards: xtensa: adsp: select IMX for imx8 board

iuliana-prodan opened the pull request:
For some platforms, like NXP's IMX8 or Mediatek's MT8195,
the size of an interrupt vector table entry is 0x1C bytes,
less than usual (0x30 for Intel's platforms).
So, the interrupt handlers don't fit in the vector table entries.
    
I've added a small indirection to bypass this size
constraint and moved the default handlers to the end
of vector table, renaming them to _Level\LVL\()VectorHelper.
For this, I've added a generic configuration -
XTENSA_SMALL_VECTOR_TABLE_ENTRY and enabled it for
NXP_ADSP_IMX8 board.

Fixes #38216
Signed-off-by: Iuliana Prodan <iuliana.prodan@nxp.com>

MaureenHelm reviewed (commented):
This symbol is defined in SOF, shouldn't it be configured there instead?

iuliana-prodan commented:
> This symbol is defined in SOF, shouldn't it be configured there instead?

It is configured in SOF, that's why the SOF application was passing, but not the kernel tests.
That was actually the problem, that this is not defined in Zephyr also.

nashif reviewed (changes requested):
> This symbol is defined in SOF, shouldn't it be configured there instead?

oh, have not noticed that. This cross Kconfig contamination should be avoided. What is CONFIG_IMX exactly? if that is supposed to indicate an SoC family, this needs to come from Zephyr, not SOF. The board, SoC,  architecture kconfigs need to be in Zephyr.

dbaluta commented:
@anasnashif CONIFG_IMX it is supposed to indicate a SoC family.

iuliana-prodan commented:
> > This symbol is defined in SOF, shouldn't it be configured there instead?
> 
> oh, have not noticed that. This cross Kconfig contamination should be avoided. What is CONFIG_IMX exactly? if that is supposed to indicate an SoC family, this needs to come from Zephyr, not SOF. The board, SoC, architecture kconfigs need to be in Zephyr.

@nashif With this patch, now CONFIG_IMX comes from Zephyr.
If we want to use nxp_adsp_imx8 board it will select the IMX family, for all tests/applications (kernel's or SOF).

We also have a CONFIX_IMX in SOF, and for subsys/audio/sof application this will be selected twice - first from Zephyr and next from SOF.

To fix this I can change the name of the CONFIG_IMX from Zephyr, I can call it CONFIG_SOC_IMX

iuliana-prodan committed:
arch: xtensa: add workaround for small vector table entries

For some platforms, like NXP's IMX8 or Mediatek's MT8195,
the size of an interrupt vector table entry is 0x1C bytes,
less than usual (0x30 for Intel's platforms).
So, the interrupt handlers don't fit in the vector table
entries.

I've added a small indirection to bypass this size
constraint and moved the default handlers to the end
of vector table, renaming them to
_Level\LVL\()VectorHelper.
For this, I've added a generic configuration -
XTENSA_SMALL_VECTOR_TABLE_ENTRY.

Signed-off-by: Iuliana Prodan <iuliana.prodan@nxp.com>

iuliana-prodan committed:
boards: xtensa: adsp: select XTENSA_SMALL_VECTOR_TABLE_ENTRY for imx8 board

Add CONFIG_XTENSA_SMALL_VECTOR_TABLE_ENTRY
for NXP_ADSP_IMX8 board to bypass the size
constraint of the interrupt vector table entry.

Fixes #38216
Signed-off-by: Iuliana Prodan <iuliana.prodan@nxp.com>

iuliana-prodan commented:
> > > This symbol is defined in SOF, shouldn't it be configured there instead?
> > 
> > 
> > oh, have not noticed that. This cross Kconfig contamination should be avoided. What is CONFIG_IMX exactly? if that is supposed to indicate an SoC family, this needs to come from Zephyr, not SOF. The board, SoC, architecture kconfigs need to be in Zephyr.
> 
> @nashif With this patch, now `CONFIG_IMX` comes from Zephyr.
> If we want to use nxp_adsp_imx8 board it will select the IMX family, for all tests/applications (kernel's or SOF).
> 
> We also have a` CONFIX_IMX` in SOF, and for `subsys/audio/sof` application this will be selected twice - first from Zephyr and next from SOF.
> 
> To fix this I can change the name of the `CONFIG_IMX` from Zephyr, I can call it `CONFIG_SOC_IMX`

@nashif @MaureenHelm Actually I've chosen another solution - more generic and it can be used by other platforms too.

So, I've add a generic CONFIG to bypass the size constraint we (and other platforms like Mediatek's [MT8195](https://github.com/thesofproject/sof/blob/main/src/platform/mt8195/include/platform/lib/memory.h#L155)) have for vector table entry - see [commit1](https://github.com/zephyrproject-rtos/zephyr/pull/38431/commits/a9b8d8d4b11216e3e74293ad517bbb8fa2d52cea), and enabled it for `nxp_adsp_imx8` board - see [commit2](https://github.com/zephyrproject-rtos/zephyr/pull/38431/commits/550ff6cd2dee1bf6855fe22974650b1d2d67b94c).

cc: @dbaluta @andyross @dcpleung

nashif commented:
> @nashif @MaureenHelm Actually I've chosen another solution - more generic and it can be used by other platforms too.

yes, this looks better. We should not have SoC related kconfigs in architecture code anyways. So this is an improvement.

MaureenHelm commented on arch/xtensa/Kconfig:
```diff
@@ -90,4 +90,13 @@ config XTENSA_WAITI_BUG
 
+config XTENSA_SMALL_VECTOR_TABLE_ENTRY
+	bool "Enable workaround for small vector table entries"
+	default n
```
You can remove this line, bools are implicitly default n.

iuliana-prodan replied:
Done!

MaureenHelm replied:
Thank you!

MaureenHelm reviewed (changes requested):
One minor nit, but otherwise I agree with @nashif this is a better solution.
